fix(forms): make composition event buffering configurable

kara opened the pull request:
This commit fixes a regression where `ngModel` no longer syncs letter by letter on Android devices, and instead syncs at the end of every word. This broke when we introduced buffering of IME events so IMEs like Pinyin keyboards or Katakana keyboards wouldn't display composition strings. Unfortunately, iOS devices and Android devices have opposite event behavior. Whereas iOS devices fire composition events for IME keyboards only, Android fires composition events for Latin-language keyboards. For this reason, languages like English don't work as expected on Android if we always buffer. So to support both platforms, composition string buffering will only be turned on by default for non-Android devices.

However, we have also added a `COMPOSITION_BUFFER_MODE` token to make this configurable by the application, because in some cases, apps might might still want to receive intermediate values. For example, some inputs begin searching behavior based on Latin letters before a character selection is made.

As a provider, this is fairly flexible. If you want to turn composition buffering off, simply provide the token at the top level:

```ts
providers: [
   {provide: COMPOSITION_BUFFER_MODE, useValue: false}
]
```

Or, if you want to change the mode  based on locale or platform, you can use a factory:

```ts
import {shouldUseBuffering} from 'my/lib';

....
providers: [
   {provide: COMPOSITION_BUFFER_MODE, useFactory: shouldUseBuffering}
]
```

Closes #15079.

robwormald commented on packages/forms/src/directives/default_value_accessor.ts:
```diff
@@ -32,15 +35,25 @@ export const DEFAULT_VALUE_ACCESSOR: any = {
-  host: {'(input)': 'onChange($event.target.value)', '(blur)': 'onTouched()'},
+  // selector: '[ngModel],[formControl],[formControlName]',
+  host: {
+    '(input)': '_handleInput($event.target.value)',
```
does the `compositionupdate` event come into this at all? Does it make sense to listen to that instead of `input` ?

kara replied:
Talked in person about this

robwormald commented:
LGTM once tests green

tinayuangao reviewed (commented):
LGTM

IgorMinar commented on tools/public_api_guard/forms/forms.d.ts:
```diff
@@ -140,7 +143,10 @@ export interface ControlValueAccessor {
     onTouched: () => void;
-    constructor(_renderer: Renderer, _elementRef: ElementRef);
+    constructor(_renderer: Renderer, _elementRef: ElementRef, _compositionMode: boolean);
+    _compositionEnd(value: any): void;
```
it's a bit weird that you decided to use _ prefix just for these methods. they are all "private" no?

kara replied:
Talked in person about this

DzmitryShylovich replied:
they must be public because they are used in event bindings. But will it work with `@internal`? just curious.

IgorMinar commented on packages/forms/test/template_integration_spec.ts:
```diff
@@ -1166,6 +1133,67 @@ export function main() {
 
     });
 
+    describe('IME events', () => {
```
based on how these tests are written, it is not clear what's the default and that the default works on both android and ios.

is it feasible to change the tests to test this?

kara committed:
fix(forms): make composition event buffering configurable

This commit fixes a regression where `ngModel` no longer syncs
letter by letter on Android devices, and instead syncs at the
end of every word. This broke when we introduced buffering of
IME events so IMEs like Pinyin keyboards or Katakana keyboards
wouldn't display composition strings. Unfortunately, iOS devices
and Android devices have opposite event behavior. Whereas iOS
devices fire composition events for IME keyboards only, Android
fires composition events for Latin-language keyboards. For
this reason, languages like English don't work as expected on
Android if we always buffer. So to support both platforms,
composition string buffering will only be turned on by default
for non-Android devices.

However, we have also added a `COMPOSITION_BUFFER_MODE` token
to make this configurable by the application. In some cases, apps
might might still want to receive intermediate values. For example,
some inputs begin searching based on Latin letters before a
character selection is made.

As a provider, this is fairly flexible. If you want to turn
composition buffering off, simply provide the token at the top
level:

```ts
providers: [
   {provide: COMPOSITION_BUFFER_MODE, useValue: false}
]
```

Or, if you want to change the mode  based on locale or platform,
you can use a factory:

```ts
import {shouldUseBuffering} from 'my/lib';

....
providers: [
   {provide: COMPOSITION_BUFFER_MODE, useFactory: shouldUseBuffering}
]
```

Closes #15079.
